kernel/src/table_changes/log_replay/tests.rs

```diff
@@ -0,0 +1,514 @@
+
+    let res = LogReplayScanner::prepare_table_changes(
+        commits.next().unwrap(),
+        engine.as_ref(),
+        &get_schema().into(),
+    );
+
+    assert!(matches!(res, Err(Error::ChangeDataFeedUnsupported(_))));
+}
+
+#[tokio::test]
+async fn unsupported_reader_feature() {
+    let engine = Arc::new(SyncEngine::new());
+    let mut mock_table = LocalMockTable::new();
+    mock_table
+        .commit([Protocol::try_new(
+            3,
+            7,
+            Some(["deletionVectors", "columnMapping"]),
+            Some([""; 0]),
+        )
+        .unwrap()
+        .into()])
+        .await;
+
+    let mut commits = get_segment(engine.as_ref(), mock_table.table_root(), 0, None)
+        .unwrap()
+        .into_iter();
+
+    let res = LogReplayScanner::prepare_table_changes(
+        commits.next().unwrap(),
+        engine.as_ref(),
+        &get_schema().into(),
+    );
+    println!("Res: {:?}", res.err());
+
+    //assert!(matches!(res, Err(Error::ChangeDataFeedUnsupported(_))));
+}
+#[tokio::test]
+async fn column_mapping_should_fail() {
+    let engine = Arc::new(SyncEngine::new());
+    let mut mock_table = LocalMockTable::new();
+    let schema_string = serde_json::to_string(&get_schema()).unwrap();
+    mock_table
+        .commit([Metadata {
+            schema_string,
+            configuration: HashMap::from([
+                (
+                    "delta.enableDeletionVectors".to_string(),
+                    "true".to_string(),
+                ),
+                ("delta.enableChangeDataFeed".to_string(), "true".to_string()),
+                ("delta.columnMapping.mode".to_string(), "id".to_string()),
+            ]),
+            ..Default::default()
+        }
+        .into()])
+        .await;
+
+    let mut commits = get_segment(engine.as_ref(), mock_table.table_root(), 0, None)
+        .unwrap()
+        .into_iter();
+
+    let res = LogReplayScanner::prepare_table_changes(
+        commits.next().unwrap(),
+        engine.as_ref(),
+        &get_schema().into(),
+    );
+
+    assert!(matches!(res, Err(Error::ChangeDataFeedUnsupported(_))));
+}
+
+#[tokio::test]
+async fn incompatible_schema() {
+    let engine = Arc::new(SyncEngine::new());
+    let mut mock_table = LocalMockTable::new();
+
+    // The original schema has two fields: `id` and value.
+    let schema = get_schema().project(&["id"]).unwrap();
+    let schema_string = serde_json::to_string(&schema).unwrap();
+    mock_table
+        .commit([Metadata {
+            schema_string,
+            configuration: HashMap::from([
+                ("delta.enableChangeDataFeed".to_string(), "true".to_string()),
+                (
+                    "delta.enableDeletionVectors".to_string(),
+                    "true".to_string(),
+                ),
+            ]),
+            ..Default::default()
+        }
+        .into()])
+        .await;
+
+    let mut commits = get_segment(engine.as_ref(), mock_table.table_root(), 0, None)
```

Comment:
nit: maybe add a comment saying commit 0 which has id and value, then the new commit 1 has just id IIUC

Reply:
also might be worth having a couple other schema checks for changing type and adding column. maybe need to think about type widening here too. sounds like we currently fail for type widening. maybe have a test for that too and we can just make note that it will likely change in the future when we support type widening

Reply:
added checks for:
- Adding non-nullable column
- adding nullable column
- commit has wider type than cdf schema
- type widening (will eventually be supported)
- cdf column is nullable while the commit schema is non-nullable (will eventually be supported)

Reply:
aweeeesome thank you!